cmd/controller-manager/app/controllermanager.go

```diff
@@ -103,9 +103,238 @@ func Run(ctx context.Context, opts *options.Options) error {
 	return nil
 }
 
+// ControllerContext defines the context object for controller
+type ControllerContext struct {
+	Mgr                         controllerruntime.Manager
+	ObjectWatcher               objectwatcher.ObjectWatcher
+	Opts                        *options.Options
+	StopChan                    <-chan struct{}
+	DynamicClientSet            dynamic.Interface
+	OverrideManager             overridemanager.OverrideManager
+	ControlPlaneInformerManager informermanager.SingleClusterInformerManager
+}
+
+// InitFunc is used to launch a particular controller.  It may run additional "should I activate checks".
+// Any error returned will cause the controller process to `Fatal`
+// The bool indicates whether the controller was enabled.
+type InitFunc func(ctx ControllerContext) (enabled bool, err error)
+
+// NewControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
+func NewControllerInitializers() map[string]InitFunc {
+	controllers := map[string]InitFunc{}
+	controllers["cluster"] = startClusterController
+	controllers["clusterStatus"] = startClusterStatusController
+	controllers["hpa"] = startHpaController
+	controllers["binding"] = startBindingController
+	controllers["execution"] = startExecutionController
+	controllers["workStatus"] = startWorkStatusController
+	controllers["namespace"] = startNamespaceController
+	controllers["serviceExport"] = startServiceExportController
+	controllers["endpointSlice"] = startEndpointSliceController
+	controllers["serviceImport"] = startServiceImportController
+	return controllers
+}
+
+func startClusterController(ctx ControllerContext) (enabled bool, err error) {
+	mgr := ctx.Mgr
+	opts := ctx.Opts
+	clusterController := &cluster.Controller{
+		Client:                    mgr.GetClient(),
+		EventRecorder:             mgr.GetEventRecorderFor(cluster.ControllerName),
+		ClusterMonitorPeriod:      opts.ClusterMonitorPeriod.Duration,
+		ClusterMonitorGracePeriod: opts.ClusterMonitorGracePeriod.Duration,
+		ClusterStartupGracePeriod: opts.ClusterStartupGracePeriod.Duration,
+	}
+	if err := clusterController.SetupWithManager(mgr); err != nil {
+		klog.Fatalf("Failed to setup cluster controller: %v", err)
```

Comment:
We don't need this since you can do this uniformly in `StartControllers`

Reply:
ok i will fix it

Reply:
How to do this by `StartControllers`? I doubt it.

Reply:
In the `StartControllers` function we need to execute these `InitFunc` one by one, so the `InitFunc` only needs to return the error:

```go
klog.V(1).Infof("Starting %q", controllerName)
started, err := initFn(ctx)
if err != nil {
	klog.Errorf("Error starting %q", controllerName)
	return err
}
```